### ISSUE TITLE
Terraform not returnning 

### ISSUE THREAD

**Issue by mbenkhal27** [vmware-archive/legacy-terraform-provider-vra7#73]
My team has been using the latest version of the plugin and We have been having  issue where the terraform keeps running for long time  after "terraform apply" and does not return, even though the machine has been provisioned successfully. Do you have any idea what is causing that to happen?

**Comment by jonnyborbs** [comment]
@virajindasrao can you please check this out? I sent you an email regarding it as well

**Comment by afurlani24** [comment]
This has happened since the last pull request to add the Call a vRA Resource Actions on a provisioned resource.  So even when the server is done in vRA the terraform command line keeps going and does not think the build is done.   I am never getting the creation complete message that I have always gotten "vra7_resource.RHEL_7_Full_Support: Creation complete (ID: 504686b1-83e3-4a14-aebc-a83cb4706137)

**Comment by virajindasrao** [comment]
May I know the `terraform show` command output after wait_timeout completion?
Also, let me know the versions of `go` and `terraform` you are using?

**Comment by afurlani24** [comment]
@vaficionado I am using Terraform version 11.7 and my version of Go is 1.9.2 I was using Terraform 11.5 and was told to try 11.7 which I did and still have the same issue.  Now even when I put a the wait_timeout it still keeps going and does not time out.  When I ran the terraform show nothing came back

**Comment by afurlani24** [comment]
@virajindasrao So I went back to Terraform 11.5 here is the output..
vra7_resource.Linux: Still creating... (32m20s elapsed)
vra7_resource.Linux: Still creating... (32m30s elapsed)
vra7_resource.Linux: Still creating... (32m40s elapsed)

Error: Error applying plan:

1 error(s) occurred:

* vra7_resource.Linux: 1 error(s) occurred:

* vra7_resource.Linux: resource is still being created

Terraform does not automatically rollback in the face of errors.
Instead, your Terraform state file has been partially updated with
any resources that successfully completed. Please address the error
above and apply again to incrementally change your infrastructure.



C:\terraformvra>terraform show
vra7_resource.Linux:
  id = 3631d83d-3353-409a-b2ca-ab80ef563005
  catalog_configuration.% = 2
  catalog_configuration.Optum.rules.use.agreement = yes
  catalog_configuration.lease_days =
  catalog_id = 83b86d57-9788-4013-934b-74d55d8c67c5
  catalog_name = RHEL 6 Full-Support
  deployment_configuration.% = 2
  deployment_configuration.description = Deployment via Terraform
  deployment_configuration.reasons = FURLANI RHEL TESTING
  request_status = IN_PROGRESS
  resource_configuration.% = 14
  resource_configuration.RHEL_6_Full_Support.Optum.AskID = UHGWM110-018404
  resource_configuration.RHEL_6_Full_Support.Optum.DR.code = DR-101
  resource_configuration.RHEL_6_Full_Support.Optum.DataCenter = CTC
  resource_configuration.RHEL_6_Full_Support.Optum.Environment = PROD
  resource_configuration.RHEL_6_Full_Support.Optum.Interface.Zone = Intranet
  resource_configuration.RHEL_6_Full_Support.Optum.Network.Zone = intranet
  resource_configuration.RHEL_6_Full_Support.Optum.ReservationPolicyName =
  resource_configuration.RHEL_6_Full_Support.Optum.ResilienceZone = None
  resource_configuration.RHEL_6_Full_Support.Optum.Server.Type = APP
  resource_configuration.RHEL_6_Full_Support.Optum.UnixID.Search = mbenkhal
  resource_configuration.RHEL_6_Full_Support.Optum.UnixID.SearchResult = mbenkhal
  resource_configuration.RHEL_6_Full_Support.VirtualMachine.Disk1 = 105
  resource_configuration.RHEL_6_Full_Support.cpu = 2
  resource_configuration.RHEL_6_Full_Support.memory = 4096
  wait_timeout = 20


C:\terraformvra>

**Comment by raghavav** [comment]
So, if I understand the problem correctly, things work fine with Terraform 11.5, but then things hang when using Terraform 11.7. Is that an accurate summary?

The code, as written, should never run beyond the timeout. The default wait_timeout is 15 min.

@virajindasrao is unable to reproduce this issue. I am going to try and repro it here in-house. So any pointers would really help.

**Comment by afurlani24** [comment]
@raghavav No it does not matter what version of Terraform I use 11.5 or 11.7 Terraform keeps going without getting a Creation complete (ID: 504686b1-83e3-4a14-aebc-a83cb4706137) message.  The server builds in vRA with no issues the state file still says In Progress.. so it is like terraform does not think the server is completed in vRA it will run until it reaches what ever time out you set.  This is happening to everyone here using the latest plugin.. if I go back to the previous binary it works but I want to be able to change CPU.. Memory on already provisioned server which that lastest Plugin is supposed to do. 
Thanks

**Comment by afurlani24** [comment]
@raghavav Let me know I could get on a webex with you to show you what is going on?
Thanks
Tony

**Comment by raghavav** [comment]
Yes, a webex would really help. In fact, could you post a webex recording right here into this issue so that other users have access to it too? Thanks for that!

And the output of 'terraform show' will also help.

**Comment by jonnyborbs** [comment]
I will connect you guys on email in a moment - standby @afurlani24 @raghavav

**Comment by raghavav** [comment]
Recording my attempts here for posterity ...

I ran through a series of 'terraform apply; terraform destroy' calls against my in-house vRA (7.4). The blueprint was a simple one with one machine in it. The provision took about 4 min.

There were no problems with running through these iterations. I was able to provision and destroy successfully, several times.

**Comment by woodyj56** [comment]
Hi Guys
I am experiencing the same issue with 0.11.7 has there been any progress on this?
Thanks

**Comment by afurlani24** [comment]
@woodyj56 Hello I just sent email to developer I have been working with to get an update they cannot reproduce it on their side I sent them a recording of what I have been doing.  Mine does not work with 11.5 or 11.7.  So does yours work with 11.5 and you are able to make post provisioning changes without issues?
Thanks

**Comment by woodyj56** [comment]
@afurlani24 thanks for responding.

actually mine doesn’t work with either as well so at the moment i cannot provision anything fully with terraform.

I’m in the early stages of testing can you or anyone recommend a stable version where i can definately expect this too work?

**Comment by raghavav** [comment]
I am still having no luck reproducing this issue. I have had several deployments go through successfully. I have tried this with properties very similar to those in the Webex recording, including additional disks. This is very strange. I am on Terraform 0.11.7, and am trying this with the latest compiled version of the plugin.

**Comment by raghavav** [comment]
Just to be sure you guys are not stumbling into an issue I was hitting with my plugin setup ...

Can you confirm that your terraform-provider-vra7 plugin is correctly installed according to these instructions: https://www.terraform.io/docs/configuration/providers.html#third-party-plugins.

I know this may be an obvious thing to check, but I did have it wrong for a brief period because I was using the wrong directory for the location of the plugin (Terraform changed the name I think).

**Comment by woodyj56** [comment]
Hi 
I hadn’t done this but i get the same result unfortunately.  It times out and fails but builds successfully on VRA.

**Comment by afurlani24** [comment]
@woodyj56 Hello you would need to get the version before this one that works fine for me.

**Comment by afurlani24** [comment]
@raghavav Hello yes I have my binary under that directory it had to be or would not pick up new plugin

**Comment by woodyj56** [comment]
@furlani24
Thanks - this is what we have done and its working fine for me now

**Comment by raghavav** [comment]
Can I have a little more of an explanation please? What works and what doesn't? Am I hearing that this works on an earlier version of Terraform? Or is it the plugin you are talking about? In my case, I have the latest version of the plugin and Terraform 0.11.7, with no issues.

**Comment by hobovirtual** [comment]
I'm getting also this behavior and decided to do more test, here's the information of my test.

I have two management box

1. Windows 10
```
Terraform v0.11.8
+ provider.vra7 (unversioned)
```

2. CentOS 7.5
```
Terraform v0.11.8
+ provider.vra7 (unversioned)
```
The user initiating the request has the User Role in the Developpers Business Group in vRealize Automation

When I launch the deployment from my Windows box, the deployment completes in vRA but i never get the response from my terraform apply and it just times out, as described above by @afurlani24 

If use the same declaration from my Linux box, it completes successfully and as expected.

------------------------

I will perform more test since when i'm using my tenantadmin account which has all the access within vRA i get the response and my terraform apply complete on my Windows box.

**Comment by afurlani24** [comment]
@hobovirtual Thank you for this information I had a webex with @raghavav yesterday to go over this issue.  This is looking more like a security role issue.  I wonder what is needed for the user role to work?
Thanks

**Comment by jonnyborbs** [comment]
Actually, this new info definitely seems to point at an OS dependent issue rather than a security role one. @afurlani24 you mentioned that you had others using Mac that were seeing the same issue, can you re-confirm that? It would definitely help us isolate the problem.

**Comment by afurlani24** [comment]
@vaficionado Yes this is also happening on my bosses Mac he is having the same result.

**Comment by raghavav** [comment]
@afurlani24 Thanks for all the logs and help with debugging this issue.
There does seem to be a bit of an assumption in the code that is causing this, that may need fixing.

Could you please check the following on VRA for me:
Is the user that is making this request entitled to the 'Reconfigure (Machine)' action under Administration -> Catalog Management -> Entitlements?
If the user is not entitled, could you please try turning on this entitlement to see if the problem goes away? If that does work, you should be able use this as a workaround for now, while I fix the core issue behind this. But please do check, could you?
Thanks
Vijay

**Comment by afurlani24** [comment]
@raghavav Hello, I will check this out this morning and get back to you.
Thanks,
Tony

**Comment by jonnyborbs** [comment]
Thanks @afurlani24 - anxiously awaiting your results

**Comment by afurlani24** [comment]
@raghavav @virajindasrao Hello so in production I do not have that Reconfigure machine entitlement and will not be able to change that in production.  I made the change in my development system and it did work!  So that is the issue that was causing that.
Thanks,
Tony

**Comment by afurlani24** [comment]
@raghavav We are entitled to Add Disk, Change CPU(Virtual machine), Change RAM(Virtual machine) not to Reconfigure Machine so this could be a problem.

**Comment by jonnyborbs** [comment]
That's actually interesting and corroborates our findings. Any chance at all you could get reconfigure access just on a test entitlement for validation purposes? Given the trouble we've had reproducing this issue it would be the (likely) last step we need to be sure we know exactly what is wrong, and then we can focus what we are going to do about it

**Comment by jonnyborbs** [comment]
Sorry, just noticed your second comment above where you validated this in your dev environment. That's very helpful, thanks.

**Comment by afurlani24** [comment]
@vaficionado Yes that is the issue it works in development and I actually made a change to CPU and it kicked off a reconfigure.

**Comment by raghavav** [comment]
Thanks @afurlani24 for confirming my theory.

You should not need this entitlement in order to provision blueprints. But you do need this entitlement to change cpu, memory etc. So, we will have to tighten the implementation a little to first allow provisioning to proceed and at the same time to clearly indicate that reconfiguration is not possible.

**Comment by afurlani24** [comment]
@raghavav Ok however I don't need that when using the UI because I have the Add Disk, Change CPU(Virtual machine), Change RAM(Virtual machine) entitlements.  So this will only be needed when using Terraform.

**Comment by raghavav** [comment]
Ah, I see @afurlani24. Maybe we are using the wrong entitlement in the plugin then. Thanks for the input. We will figure this out here.

**Comment by hobovirtual** [comment]
> I'm getting also this behavior and decided to do more test, here's the information of my test.
> 
> I have two management box
> 
> 1. Windows 10
> 
> ```
> Terraform v0.11.8
> + provider.vra7 (unversioned)
> ```
> 1. CentOS 7.5
> 
> ```
> Terraform v0.11.8
> + provider.vra7 (unversioned)
> ```
> The user initiating the request has the User Role in the Developpers Business Group in vRealize Automation
> 
> When I launch the deployment from my Windows box, the deployment completes in vRA but i never get the response from my terraform apply and it just times out, as described above by @afurlani24
> 
> If use the same declaration from my Linux box, it completes successfully and as expected.
> 
> I will perform more test since when i'm using my tenantadmin account which has all the access within vRA i get the response and my terraform apply complete on my Windows box.

FYI, updated my provider with latest version and the behavior mentioned in this quoted text, is no longer observed.

**Comment by automationbetauser** [comment]
@hobovirtual so the latest provider build has fixed the issue for you? Just to be clear

**Comment by jonnyborbs** [comment]
@hobovirtual so the latest provider build has fixed the issue for you? Just to be clear

**Comment by Prativa20** [comment]
Fix for #82 has been merged and it should fix a part of this issue which is terraform apply hangs even when the deployment is completed on the vRA.

The other part is as mentioned in the below comment. And we are working to fix it.

> @afurlani24 Thanks for all the logs and help with debugging this issue.
> There does seem to be a bit of an assumption in the code that is causing this, that may need fixing.
> 
> Could you please check the following on VRA for me:
> Is the user that is making this request entitled to the 'Reconfigure (Machine)' action under Administration -> Catalog Management -> Entitlements?
> If the user is not entitled, could you please try turning on this entitlement to see if the problem goes away? If that does work, you should be able use this as a workaround for now, while I fix the core issue behind this. But please do check, could you?
> Thanks
> Vijay

**Comment by hobovirtual** [comment]
Hi @vaficionado 

Yes exactly, after updating to latest release the issue is gone from my side.
